### DISCUSSION TITLE
Request headers validation

### DISCUSSION THREAD

**Discussion by Ale-Cas** [fastapi/fastapi#11157] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel


class ItemRequestBody(BaseModel):
    """Request body for the Suitability Profiler."""

    item_id: int
    item_name: str


app = FastAPI()


@app.post("/items/")
def create_item(item: ItemRequestBody):
    """Mock create item endpoint."""
    return item


client = TestClient(app)


def test_create_item() -> None:
    """Test create item endpoint."""
    response = client.post(
        "/items/",
        json={"item_id": 1, "item_name": "test"},
        headers={"Content-Type": "text/xml"},
    )
    assert response.status_code == 422
    assert response.json() == {
        "detail": [
            {
                "loc": ["body"],
                "msg": "value is not a valid dict",
                "type": "type_error.dict",
            }
        ]
    }
```

### Description

Hi all, 
I have a feature request, to replace a workaround that I use based on this answer https://github.com/tiangolo/fastapi/issues/2915.
I would like to have better tooling to configure validators on the request headers. 
For example, I think it would be helpful to be able to have something that looks like: 
```python
from fastapi import FastAPI
from fastapi.headers import HeadersConfig

app = FastAPI(headers_config=HeadersConfig(accepted_content_type=“application/json”))
```
So that in the example above, instead of returning the Pydantic validation (which is very generic and does not point to the root issue), FastAPI validates the content-type in the header and returns a nicer, readable message to the user, something that sounds like:
`"Content-Type": "text/xml" is not supported, please use "application/json"` 
I think it's a quite common scenario to validate and filter the request headers, and right now FastAPI gives the developer the tools to handle these validations (which is great), but I think it would be even better to integrate this in the library so that different users don't have to re-implement this from scratch. 

### Operating System

macOS

### FastAPI Version

0.86.0

### Pydantic Version

1.8.2

### Python Version

3.8.15

**Comment by ahmedabdou14** [comment]
In my opinion a middleware on the main app router is more than enough for your case.

**Comment by YuriiMotov** [comment, marked as answer]
You can do it using middleware:
```
@app.middleware("http")
async def check_content_type(request: Request, call_next):
    content_type = request.headers.get("Content-Type")
    if (content_type is not None) and (content_type != "application/json"):
        return Response(
            content=(
                f"'Content-Type': '{request.headers['Content-Type']}' "
                f"is not supported, please use 'application/json'"
            ),
            status_code=415,
        )
    response = await call_next(request)
    return response
```

And use it as follows:
```
from pydantic import BaseModel
from fastapi import FastAPI, Request, Response
from fastapi.testclient import TestClient

app = FastAPI()

@app.middleware("http")
async def check_content_type(request: Request, call_next):
    content_type = request.headers.get("Content-Type")
    if (content_type is not None) and (content_type != "application/json"):
        return Response(
            content=(
                f"'Content-Type': '{request.headers['Content-Type']}' "
                f"is not supported, please use 'application/json'"
            ),
            status_code=415,
        )
    response = await call_next(request)
    return response

class ItemRequestBody(BaseModel):
    item_id: int
    item_name: str

@app.post("/items/")
def create_item(item: ItemRequestBody):
    """Mock create item endpoint."""
    return item

client = TestClient(app)

def test_create_item_wrong_content_type() -> None:
    response = client.post(
        "/items/",
        json={"item_id": 1, "item_name": "test"},
        headers={"Content-Type": "text/xml"},
    )
    assert response.status_code == 415
    assert "is not supported, please use 'application/json'" in response.text

def test_create_item_success() -> None:
    response = client.post(
        "/items/",
        json={"item_id": 1, "item_name": "test"},
    )
    assert response.status_code == 200

def test_create_item_explicit_content_type_success() -> None:
    response = client.post(
        "/items/",
        json={"item_id": 1, "item_name": "test"},
        headers={"Content-Type": "application/json"},
    )
    assert response.status_code == 200
```
